alpa/pipeline_parallel/cross_mesh_resharding.py

```diff
@@ -1179,3 +1491,433 @@ def _args_between(src_stage, dst_stage):
+
+        device_to_worker_id_map = {}
+        worker_id_to_devices = {}
+        n_workers = 0
+        for idx, device in enumerate(self.all_devices):
+            if tmp_device_to_worker_id_map[device] == idx:
+                device_to_worker_id_map[device] = n_workers
+                worker_id_to_devices[n_workers] = [device]
+                n_workers += 1
+            else:
+                group_head_device = self.all_devices[
+                    tmp_device_to_worker_id_map[device]]
+                worker_id = device_to_worker_id_map[group_head_device]
+                device_to_worker_id_map[device] = worker_id
+                worker_id_to_devices[worker_id].append(device)
+
+        abstract_works = []
+        for work in self.works:
+            sender_ids = set()
+            for sender in work.senders:
+                sender_ids.add(device_to_worker_id_map[sender])
+            sender_ids = list(sender_ids)
+            sender_ids.sort()
+            receiver_ids = set()
+            for receiver in work.receivers:
+                receiver_ids.add(device_to_worker_id_map[receiver])
+            receiver_ids = list(receiver_ids)
+            receiver_ids.sort()
+            time_spent = work.data_size
+
+            abstract_works.append(
+                SingleAbstractedLoadBalancingWork(sender_ids, receiver_ids,
+                                                  time_spent))
+
+        if global_config.resharding_loadbalance_mode == "loadbalance_size":
+            task = AbstractedLoadBalancingOverSizeTaskSolver(
+                n_workers, abstract_works)
+        else:
+            if global_config.loadbalance_order_algo == "search":
+                task = AbstractedLoadBalancingTaskSolverSearchAlgo(
+                    n_workers, abstract_works)
+            else:
+                task = AbstractedLoadBalancingTaskSolverGreedyAlgo(
+                    n_workers, abstract_works)
+
+        sol_assigned_sender_id, sol_order = task.solve()
+
+        sol_assigned_sender = []
+        for work, worker_id in zip(self.works, sol_assigned_sender_id):
+            selected_sender = None
+            for sender in work.senders:
+                if device_to_worker_id_map[sender] == worker_id:
+                    selected_sender = sender
+                    break
+            assert selected_sender is not None
+            sol_assigned_sender.append(selected_sender)
+        return sol_assigned_sender, sol_order
+
+    def print_task(self):
+        print("\nTask[START]")
+        print(f"src_mesh_devices: {self.src_mesh_devices}")
+        print(f"dst_mesh_devices: {self.dst_mesh_devices}")
+        print(f"device_host_map: {self.device_host_map}")
+        print("works:")
+        for work in self.works:
+            print(work)
+        print("nic_contraints:")
+        for contraint in self.nic_contraints:
+            print(contraint)
+        print("Task[END]\n")
+
+
+class AbstractedLoadBalancingTaskSolver:
```

Comment:
use `@abstractmethod` for methods in the abstracted class, and for non-abstracted class, please do not use `Abstracted` in their names

Reply:
OK